go-controller/pkg/clustermanager/networkconnect/cluster_network_connect.go

```diff
@@ -0,0 +1,305 @@
+		if err != nil {
+			return err
+		}
+		err = util.UpdateNetworkConnectRouterTunnelKeyAnnotation(cnc.Name, c.cncClient, tunnelID[0])
+		if err != nil {
+			return err
+		}
+		cncState.tunnelID = tunnelID[0]
+	}
+	// STEP3: Discover the selected UDNs and CUDNs
+	discoveredNetworks, allMatchingNADKeys, err := c.discoverSelectedNetworks(cnc)
+	if err != nil {
+		return err
+	}
+	// STEP4: Generate or release subnets of size CNC.Spec.ConnectSubnets.NetworkPrefix for each layer3 network
+	//  and /31 or /127 subnets for each layer2 network
+	// We intentionally don't compute or use the networksNeedingAllocation set here because we want to return all
+	// currently allocated subnets for each owner back to the annotation update step.
+	allocatedSubnets, allMatchingNetworkKeys, err := c.allocateSubnets(discoveredNetworks, cncState.allocator)
+	if err != nil {
+		return err
+	}
+	// This step will handle the release of subnets for networks that are no longer matched or are deleted.
+	networksNeedingRelease := cncState.selectedNetworks.Difference(allMatchingNetworkKeys)
+	if len(networksNeedingRelease) > 0 {
+		err = c.releaseSubnets(networksNeedingRelease, cncState.allocator)
+		if err != nil {
+			return err
+		}
+	}
+	networksNeedingAllocation := allMatchingNetworkKeys.Difference(cncState.selectedNetworks)
+	klog.V(5).Infof("DEBUG: cncState.selectedNetworks=%v, allMatchingNetworkKeys=%v, networksNeedingAllocation=%v, networksNeedingRelease=%v",
+		cncState.selectedNetworks.UnsortedList(), allMatchingNetworkKeys.UnsortedList(), networksNeedingAllocation.UnsortedList(),
+		networksNeedingRelease.UnsortedList())
+	// we need to update the annotation only if there are networks that are newly matched or newly released
+	if len(networksNeedingAllocation) > 0 || len(networksNeedingRelease) > 0 {
+		err = util.UpdateNetworkConnectSubnetAnnotation(cnc, c.cncClient, allocatedSubnets)
+		if err != nil {
+			return err
+		}
+	}
+	// plumbing is now done, update the cache with latest
+	cncState.selectedNADs = allMatchingNADKeys
+	klog.V(5).Infof("Updated selectedNADs cache for CNC %s with %d NADs", cncName, allMatchingNADKeys.Len())
+	cncState.selectedNetworks = allMatchingNetworkKeys
+	klog.V(5).Infof("Updated selectedNetworks cache for CNC %s with %d networks", cncName, allMatchingNetworkKeys.Len())
+	return nil
+}
+
+func (c *Controller) discoverSelectedNetworks(cnc *networkconnectv1.ClusterNetworkConnect) ([]*util.NetInfo, sets.Set[string], error) {
+	discoveredNetworks := []*util.NetInfo{}
+	allMatchingNADKeys := sets.New[string]()
+	for _, selector := range cnc.Spec.NetworkSelectors {
+		switch selector.NetworkSelectionType {
+		case apitypes.ClusterUserDefinedNetworks:
+			networkSelector, err := metav1.LabelSelectorAsSelector(&selector.ClusterUserDefinedNetworkSelector.NetworkSelector)
+			if err != nil {
+				return nil, nil, err
+			}
+			nads, err := c.nadLister.List(networkSelector)
+			if err != nil {
+				return nil, nil, err
```

Comment:
I am also thinking if we should continue with the other networks handling here (and in all the other returns) instead of just returning an error
otherwise one wrong NAD may break the whole CNC

Reply:
this is done - ive also added a test for it.

Reply:
spoke with @npinaeva - we agree that's the expected behaviour and for an already allocated network - it should be a very very rare event that the NAD Parse OR lister or fetch etc doesn't work and we end up releasing and re-allocating accidentally causing disruption..

The gain we get from continuing to process the good networks v/s ignoring 1 bad network is much more - so that's the right thing to do here and this was also called out as a must behaviour in the enhancement.